[RFC] Tagged type (#733)
graphql/graphql-spec: spec/Section 3 -- Type System.md

`Object` types, which define a set of fields, where each field is another
type in the system, allowing the definition of arbitrary type hierarchies.

> GraphQL supports two abstract types: interfaces, and unions.

Author: That's exactly the conclusion that I came to; I see when I edited it back out that I left a comma in there :laughing:

type system claims a union will be returned, one of the possible types will be
returned.

> A `Tagged Type` defines a list of possible members; whenever the type system

Author: I've addressed this, but I'm not happy with the wording so I'm not marking this conversation resolved.

}
```

A field of an Object type may be a Scalar, Enum, another Object type, an

> Interface, a Union or a Tagged type. Additionally, it may be any wrapping type

Reviewer: Nit: Before, you used the order:

> object types, tagged types, interfaces, and unions.

}
```

The type of an object field argument must be an input type (any type for which

> {IsInputType(type)} returns true).

Author: Done :+1:

objects and interfaces can then implement these interfaces which requires that
the implementing type will define all fields defined by those interfaces.

Fields on a GraphQL interface have the same rules as fields on a GraphQL
object; their type can be Scalar, Object, Enum, Interface, Union, Tagged types
where {IsOutputType(type)} returns {true}, or any wrapping type whose base type

> is one of those six.

Author: Done :+1:

1. A Union type must include one or more unique member types.

> 2. The member types of a Union type must all be Object base types;

Reviewer 2: Given this and other conversations around "concrete" types (e.g. re `__typename`) makes me wonder if we should RFC the idea of a "concrete" type into the spec everywhere as a prerequisite before we add a tagged type?

> 2. The member types of a Union type must all be Object base types;

Author: Initially I was hesitant to add Tagged to Unions; but I like this idea @leebyron, particularly because it solves the "union of objects and scalars" problem in a relatively clean way:

```graphql
type City {
  id: ID!
  name: String!
  coordinates: Coordinates!
}

type Town {
  id: ID!
  name: String!
  coordinates: Coordinates!
}

type POI {
  id: ID!
  name: String!
  coordinates: Coordinates!
  nearestCity: City!
}

type Coordinates {
  latitude: Float!
  longitude: Float!
}

tagged LocationSpecifier {
  locationName: String!
  locationCoordinates: Coordinates!
}

union Place = City | Town | POI | LocationSpecifier;

type Query {
  myFavoritePlace: Place
}
```

Now your favourite place could be a known city, town, point of interest, a custom "named place", or a custom position on the globe.

Let's leave it for a follow up PR; I don't _think_ that it would need any breaking changes... Do you?

1. The named type must already be defined and must be a Union type.
2. The member types of a Union type extension must all be Object base types;
   Scalar, Interface, Union and Tagged types must not be member types of a
   Union.  Similarly, wrapping types must not be member types of a Union.
3. All member types of a Union type extension must be unique.
4. All member types of a Union type extension must not already be a member of
   the original Union type.
5. Any non-repeatable directives provided must not already apply to the
   original Union type.

## Tagged Types

TaggedTypeDefinition : Description? tagged Name Directives[Const]? TaggedMembersDefinition?

TaggedTypeExtension :
  - extend tagged Name Directives[Const]? TaggedMembersDefinition
  - extend tagged Name Directives[Const]

TaggedMembersDefinition : { TaggedMemberDefinition+ }

TaggedMemberDefinition : Description? Name : Type Directives[Const]?

Tagged types represent a list of possible named members, exactly one of which
must be present. This resulting member must yield a value of a specific type.

> In outputs, like for Objects, Tagged types should be serialized as ordered

Reviewer: I don't see why ordering plays a role here, since there is only ever one field. Of course, the fields of the parent Object type in which the Tagged type is used must be ordered - but that is a concern of the Object type.

Author: (For others, it's because `__typename` and aliases can also exist.)

maps: the queried member names (or aliases) that are present are the keys and
the result of evaluating the member is the value, ordered by the order in which
they appear in the query. Note that Tagged types may be objects with more than
one key even though they guarantee that exactly one member is present, this is
because aliases may be used, and `__typename` may also be queried (and will
return the name of the tagged type).

All members defined within a Tagged type must not have a name which begins with
{"__"} (two underscores), as this is used exclusively by GraphQL's
introspection system.

> For example, a type `StringFilter` could be described as:

Reviewer: There might be an example that makes the intent of allowing only one of the fields more immediately clear. Here, passing both `lengthAtLeast` and `lengthAtMost` seems reasonable.

How about something like this:

```graphql
tagged Angle {
  degree: Float!
  radians: Float!
}
```

Author: I definitely see your point. I'll try and think of a mutually-exclusive example which uses mixed scalars.

Reviewer: Yeah, the mixed scalars illustrate the point better.

Author: I've added a type:

```graphql example
tagged PackQuantity {
  volumeInMetresCubed: Float!
  numberOfBoxes: Int!
  description: String!
}
```

I might be moving house soon :wink:

  lengthAtMost
}
```

Could yield one of the following objects:

- `{ "contains": "Awesome" }`
- `{ "lengthAtLeast": 3 }`
- `{ "lengthAtMost": 42 }`

Valid queries must supply a nested field set for a field that returns a Tagged
type, so for this schema:

```graphql example
type Query {
  stringFilter: StringFilter
}
```

This query is not valid:

```graphql counter-example
{
  stringFilter
}
```

However, this query is valid:

```graphql example
{
  stringFilter {
    contains
  }
}
```

And may yield one of the following objects:

- `{ "contains": "Awesome" }`
- `{}`

**Field Ordering**

Like with Object types, when querying a Tagged type, the resulting mapping of
fields are conceptually ordered in the same order in which they were
encountered during query execution, excluding fragments for which the type does
not apply and fields or fragments that are skipped via `@skip` or `@include`
directives, and fields which are neither the single matched member nor the
`__typename` introspection field. This ordering is correctly produced when
using the {CollectFields()} algorithm.

**Result Coercion**

Determining the result of coercing an object or tagged type is the heart of the

> GraphQL executor, so this is covered in that section of the spec. Note that

Reviewer: Can we link to "that section"?

Author: This is effectively a copy/paste of another part of the spec that has very similar wording. If we fix it here we should fix it there as well.

> GraphQL executor, so this is covered in that section of the spec. Note that

Reviewer: How about we move that note to a separate line that starts the paragraph to match the `**Input coercion**` section right below it?

only Tagged types for which {IsOutputType(type)} returns {true} may be used in
output.

**Input Coercion**

Only Tagged types for which {IsInputType(type)} returns {true} may be used as
input.

> The value for an input Tagged type should be an input object literal or an

Reviewer: ```suggestion
The value for an input Tagged type must be an input object literal or an
```

Author: This "should" is the same word used in Input Coercion for Input Objects; I think the `should/otherwise ... must` is unambiguous, so I'm going to leave this as is, but I appreciate you flagging it :+1:

unordered map supplied by a variable, otherwise a query error must be thrown.
In either case, the input object literal or unordered map must not contain any
entries with names not defined by a member of this Tagged type, otherwise an

> error must be thrown. Similarly the input object literal or unordered map must

Reviewer: ```suggestion
error must be thrown. The input object literal or unordered map must
```

contain exactly one member, otherwise an error must be thrown.

The result of coercion is an unordered map with an entry for exactly one
member both defined by the tagged type and present in the input. The resulting
map is constructed with the following rules:

> * If more than one field is defined in the input, an error should be thrown.

Reviewer: Above it says must for the same condition.

* For the single field defined in the input, if the value is {null} and the

>   field's type is a non-null type, an error should be throw.

Reviewer: ```suggestion
  field's type is a non-null type, an error must be thrown.
```

`{ a: "abc", b: "123" }` | `{}`                    | Error: Exactly one key must be specified
`{ b: "123" }`           | `{}`                    | Error: Incorrect value for member {b}
`{ a: "abc" }`           | `{}`                    | `{ a: "abc" }`
`{ b: $var }`            | `{}`                    | Error: No keys were specified
`$var`                   | `{ var: { a: "abc" } }` | `{ a: "abc" }`
`{ a: "abc", b: null }`  | `{}`                    | Error: Exactly one key must be specified
`{ b: $var }`            | `{ var: null }`         | Error: {b} must be non-null.
`{ b: 123, c: "xyz" }`   | `{}`                    | Error: Unexpected member {c}

**Type Validation**

Tagged types have the potential to be invalid if incorrectly defined. This set
of rules must be adhered to by every Tagged type in a GraphQL schema.

1. A Tagged type must define one or more members.
2. For each member of a Tagged type:
   1. The member must have a unique name within that Tagged type;
      no two members may share the same name.
   2. The member must not have a name which begins with the
      characters {"__"} (two underscores).

> 3. {IsOutputType(fieldType)} and {IsInputType(type)} cannot both be {false}.

Reviewer: What do `fieldType` and `type` refer to? This part may be expanded a bit for clarity, without reading the rest of the document it is not immediately clear what it means.

Author: They should both be `taggedType`; good spot. The "Objects" section references `objectType` without further qualification; so I think this is fine (now it's fixed).

1. The named type must already be defined and must be a Union type.
2. The member types of a Union type extension must all be Object base types;
   Scalar, Interface, Union and Tagged types must not be member types of a
   Union.  Similarly, wrapping types must not be member types of a Union.
3. All member types of a Union type extension must be unique.
4. All member types of a Union type extension must not already be a member of
   the original Union type.
5. Any non-repeatable directives provided must not already apply to the
   original Union type.

## Tagged Types

TaggedTypeDefinition : Description? tagged Name Directives[Const]? TaggedMembersDefinition?

TaggedTypeExtension :
  - extend tagged Name Directives[Const]? TaggedMembersDefinition
  - extend tagged Name Directives[Const]

TaggedMembersDefinition : { TaggedMemberDefinition+ }

TaggedMemberDefinition : Description? Name : Type Directives[Const]?

> Tagged types represent a list of possible named members, exactly one of which

Reviewer 2: Curious for thoughts on the use of a new term "members" here vs just reusing "fields" again. "field" already has some ambiguous usage when used across input and output object types, so I can see both sides of the argument to reuse it or invent a new term.

Author: I think `member fields` makes most sense. They're field-like except that exactly one will be present so the nullability rules you know about input fields, for example, don't apply here. (A tagged input type could have 3 non-nullable fields, but you'd still only specify one of them.)

Author: I've moved to using `member fields` so am resolving this :+1:

return the name of the tagged type).

All members defined within a Tagged type must not have a name which begins with
{"__"} (two underscores), as this is used exclusively by GraphQL's
introspection system.

For example, a type `StringFilter` could be described as:

```graphql example
tagged StringFilter {
  startsWith: String!
  contains: String!
  lengthAtLeast: Int!
}
```

In this case we're representing exactly one of the following:

- an object containing a single key `startsWith` with an {String} value
- an object containing a single key `contains` with a {String} value
- an object containing a single key `lengthAtLeast` with an {Int} value

The `StringFilter` Tagged type is valid as both an input and output type, but
this is not true of all tagged types. If a Tagged type has a member whose type
is only valid for output, then the Tagged type is only valid for output. If a
Tagged type has a member whose type is only valid for input, then the Tagged
type is only valid for input. If a Tagged type has a member that is only valid
for input, and another member that's only valid for output, then it is not a
valid Tagged type.

> A member of a Tagged type may be a Scalar, Enum, Input Object type, Object

Reviewer 2: Is this list just literally any type? If not it may be clearer to list what is excluded, not what is included.

Author: It's all of the currently defined types, yes, but I worried that stating "any type" might limit future option value. You're right though, clarity wins. Fixed :+1:

type, an Interface, a Union or another Tagged type. Additionally, it may be
any wrapping type (e.g. list, non-null, or any combination thereof) whose
underlying base type is one of those six.

Selecting all the members of our `StringFilter` type:

```graphql example
{
  startsWith
  contains
  lengthAtLeast
}
```

> Could yield one of the following objects:

Reviewer 3: s/objects/results

Author: Done :+1:

- `{ "contains": "awesome" }`
- `{ "lengthAtLeast": 3 }`

Valid queries must supply a nested field set for a field that returns a Tagged
type, so for this schema:

```graphql example
type Query {
  stringFilter: StringFilter
}
```

This query is not valid:

```graphql counter-example
{
  stringFilter
}
```

However, this query is valid:

```graphql example
{
  stringFilter {
    contains
  }
}
```

And may yield one of the following objects:

- `{ "contains": "awesome" }`
- `{}`

**Field Ordering**

Like with Object types, when querying a Tagged type, the resulting mapping of
fields are conceptually ordered in the same order in which they were
encountered during query execution, excluding fragments for which the type does
not apply and fields or fragments that are skipped via `@skip` or `@include`
directives, and fields which are neither the single matched member nor the
`__typename` introspection field. This ordering is correctly produced when
using the {CollectFields()} algorithm.

**Result Coercion**

Determining the result of coercing an object or tagged type is the heart of the
GraphQL executor, so this is covered in that section of the spec.

> Note only Tagged types for which {IsOutputType(type)} returns {true} may be

Reviewer 3: colon after `Note:`

Author: Done :+1:

`{ a: null, b: 123 }`    | `{}`                    | Error: Exactly one key must be specified
`{ b: 123 }`             | `{}`                    | `{ b: 123 }`
`{ a: $var, b: 123 }`    | `{ var: null }`         | Error: Exactly one key must be specified
`{ a: $var, b: 123 }`    | `{}`                    | `{ b: 123 }`
`{ b: $var }`            | `{ var: 123 }`          | `{ b: 123 }`
`$var`                   | `{ var: { b: 123 } }`   | `{ b: 123 }`
`"abc123"`               | `{}`                    | Error: Incorrect value
`$var`                   | `{ var: "abc123" } }`   | Error: Incorrect value
`{ a: "abc", b: "123" }` | `{}`                    | Error: Exactly one key must be specified
`{ b: "123" }`           | `{}`                    | Error: Incorrect value for member {b}
`{ a: "abc" }`           | `{}`                    | `{ a: "abc" }`
`{ b: $var }`            | `{}`                    | Error: No keys were specified
`$var`                   | `{ var: { a: "abc" } }` | `{ a: "abc" }`
`{ a: "abc", b: null }`  | `{}`                    | Error: Exactly one key must be specified
`{ b: $var }`            | `{ var: null }`         | Error: {b} must be non-null.

> `{ b: 123, c: "xyz" }`   | `{}`                    | Error: Unexpected member {c}

Reviewer 3: this example implies that an implementation would check that each member field is valid before checking that only one field was provided, I'm not confident that's what we should suggest

Author: Good catch; that was unintended :+1:

Author: I'm changing this to `Error: Exactly one key must be specified`

> 

Reviewer 3: nit: too many line breaks between subsections

Author: Where's prettier when you need it? :wink:

**Type Validation**

Tagged types have the potential to be invalid if incorrectly defined. This set
of rules must be adhered to by every Tagged type in a GraphQL schema.

1. A Tagged type must define one or more members.
2. For each member of a Tagged type:
   1. The member must have a unique name within that Tagged type;
      no two members may share the same name.
   2. The member must not have a name which begins with the
      characters {"__"} (two underscores).

> 3. {IsOutputType(taggedType)} and {IsInputType(taggedType)} cannot both be {false}.

Reviewer 3: <3

### Member Deprecation

Members in a Tagged type may be marked as deprecated as deemed necessary by
the application. It is still legal to query for these members on outputs or
supply these members in inputs (to ensure existing clients are not broken by
the change), but the members should be appropriately treated in documentation
and tooling.

When using the type system definition language, `@deprecated` directives are
used to indicate that a field is deprecated:

```graphql example
tagged ExampleType {

>   oldField: String @deprecated

Reviewer 3: thought: the `@deprecated` directive needs its locations updated to allow on tagged members

Author: Done :+1:

}
```

Note: a deprecated member of a tagged type may legitimately be marked as

> non-nullable since this member need not be supplied/returned.

Reviewer 3: Not sure I understand this

Author: Yeah; not sure how to word that!

For the following tagged type:

```graphql
tagged Foo {
  a: Int!
  b: Int! @deprecated(reason: "Example")
  c: Int!
}
```

Input `{a:1}` is still valid - `b`'s nullability is irrelevant. There's no benefit to enforcing `b` be nullable. If `{b:2}` is supplied it's still valid (despite being deprecated); there's no need to allow `{b: null}` which is (probably) undesirable input.

For output, selection set `{ a b c }` is still valid, and the server can simply choose to never return `b` as a result, so b's nullability is irrelevant.

Is that any clearer? How can I make this understandable. (Namely the behaviour of @deprecated w.r.t nullability differs to the rest of the spec.)

Author: I've pushed a slightly reworded note.

  - extend tagged Name Directives[Const]

Tagged type extensions are used to represent a Tagged type which has been
extended from some original Tagged type. For example, this might be used to
represent a GraphQL service which is itself an extension of another GraphQL
service.

In this example, an additional option `startsWith` is added to a `StringFilter`
type:

```graphql example
extend tagged StringFilter {
  startsWith: String!
}
```

Tagged type extensions may choose not to add additional members, instead only
adding directives.

In this example, a directive is added to a `StringFilter` type without adding
members:

```graphql example
extend tagged StringFilter @addedDirective
```

**Type Validation**

Tagged type extensions have the potential to be invalid if incorrectly defined.

1. The named type must already be defined and must be a Tagged type.
2. The members of a Tagged type extension must have unique names; no two members
   may share the same name.
3. The members of a Tagged type extension must not have names beginning with
   {"__"}.
4. Any members of a Tagged type extension must not be already defined on the
   original Tagged type.

> 5. Any non-repeatable directives provided must not already apply to the

Reviewer 3: You might want to also repeat the rule that after adding new members, IsInputType and IsOutputType can't both be false

Author: :+1:

that is only valid for input, and another member field that's only valid for
output, then it is not a valid Tagged type.

A member field of a Tagged type may be of any valid GraphQL type, including
Scalar, Enum, Input Object type, Object type, an Interface, a Union, another
Tagged type, or any wrapping type (e.g. list, non-null, or any combination
thereof) whose underlying base type is one of those seven.

Selecting all the member fields of our `PackQuantity` type:

```graphql example
{
  volumeInMetresCubed
  numberOfBoxes
  description
}
```

Could yield one of the following results:

- `{ "volumeInMetresCubed": 3.8 }`
- `{ "numberOfBoxes": 27 }`
- `{ "description": "4 large boxes, 7 medium boxes, and 16 small boxes" }`

Valid queries must supply a nested field set for a field that returns a Tagged
type, so for this schema:

```graphql example
type Query {
  packQuantity: PackQuantity
}
```

This query is not valid:

```graphql counter-example
{
  packQuantity
}
```

However, this query is valid:

```graphql example
{
  packQuantity {
    numberOfBoxes
  }
}
```

And may yield one of the following results:

- `{ "numberOfBoxes": 27 }`

> - `{}`

Reviewer 2: Permitting empty return objects is one of the parts of the output side of tagged types that makes me uncomfortable - although I can't articulate the exact rule, it feels in violation of some GraphQL principle somewhere. It's also somewhat inconsistent with how it's used on input, where providing 0 fields to a tagged type is invalid.

Because of this I am leaning more strongly towards making this an input-only type to start with.

Author: We already permit empty return objects:

```graphql
{
  __typename @skip(if: true)
}
```

Author: (We also allow it with fragment spreads; e.g. [this SWAPI example](https://graphql.org/swapi-graphql/?query=%7B%0A%20%20node(id%3A%22cGVvcGxlOjI%3D%22)%20%7B%0A%20%20%20%20...%20on%20Species%20%7B%0A%20%20%20%20%20%20name%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D)

```graphql
{
  node(id:"cGVvcGxlOjI=") {
    ... on Species {
      name
    }
  }
}
```
)

using the {CollectFields()} algorithm.

**Result Coercion**

Determining the result of coercing an object or tagged type is the heart of the
GraphQL executor, so this is covered in that section of the spec.

Note: only Tagged types for which {IsOutputType(type)} returns {true} may be
used in output.

**Input Coercion**

Only Tagged types for which {IsInputType(type)} returns {true} may be used as
input.

The value for an input Tagged type should be an input object literal or an
unordered map supplied by a variable, otherwise a query error must be thrown.
In either case, the input object literal or unordered map must not contain any
entries with names not defined by a member field of this Tagged type, otherwise
an error must be thrown. The input object literal or unordered map must contain
exactly one member field, otherwise an error must be thrown.

The result of coercion is an unordered map with an entry for exactly one member
field both defined by the tagged type and present in the input. The resulting
map is constructed with the following rules:

* If more than one field is defined in the input, an error must be thrown.

> * For the single field defined in the input, if the value is {null} and the

Reviewer 2: This rule seems like it ought to be covered by a general "the input to that single field must match the fields' type"?

Fields may accept arguments to configure their behavior. These inputs are often
scalars or enums, but they sometimes need to represent more complex values.

A GraphQL Input Object defines a set of input fields; the input fields are

> either scalars, enums, other input objects, or Tagged types for which

Author: Good spot, eagle eyes!
